src/weights.jl

```diff
@@ -35,7 +36,33 @@ end
     W(v, sum(v))
 end
 
-Base.getindex(wv::W, ::Colon) where {W <: AbstractWeights} = W(copy(wv.values), sum(wv))
+Base.getindex(wv::AbstractWeights, ::Colon) = copy(wv)
+
+Base.copy(wv::W) where {W <: AbstractWeights} =
+    weightstype(W)(copy(wv.values), sum(wv))
+
+@propagate_inbounds function Base.view(wv::W, inds...) where {S <: Real, W <: AbstractWeights{S}}
+    @boundscheck checkbounds(wv, inds...)
+    @inbounds v = invoke(view, Tuple{AbstractArray, Vararg{Any}}, wv, inds...)
+    # Sum is not actually used but compute the right type for clarity
+    weightstype(W)(v, zero(S))
+end
+
+# This method is implemented for backward compatibility
+@propagate_inbounds function Base.view(wv::W, inds::Union{Integer, CartesianIndex}...) where
+    {S <: Real, W <: AbstractWeights{S}}
+    @boundscheck checkbounds(wv, inds...)
+    @inbounds invoke(view, Tuple{AbstractArray, Vararg{Any}}, wv, inds...)
+end
+
+# Always recompute the sum for views of AbstractWeights, as we cannot know whether
+# the parent array has been mutated
+Base.sum(wv::AbstractWeights{S, T, <:SubArray{T, <:Any, <:AbstractWeights}}) where
```

Comment:
Are we sure that in the code `:sum` field is never accessed directly, but always via the `sum` function?

Reply:
No. :-D But at least it's private.

For additional safety, I considered storing `NaN`, `missing` or `nothing` in the `sum` field. But that would affect the `S` type parameter, which has no practical consequences but could be a bit weird for users (as `sum` would return a different type). Not sure.

At least by storing a zero sum, results should be obviously wrong if people access the field directly. For example, `mean` returns `Inf` and `var` returns `NaN` when the sum is zero. But I agree that's not 100% safe.

Reply:
We could also override `getproperty` for views of weights if we really wanted to prevent any incorrect behaviors. That would require using `getfield` internally in some cases.

Reply:
we can leave things as they are but I still think https://github.com/JuliaStats/StatsBase.jl/pull/723#discussion_r745115242 is a problem

Reply:
I've added a commit which uses `missing` instead of zero for views. That way any invalid access to the `sum` field will be caught (including invalid equality tests). It turns out it doesn't make the code more complex.